# Project Asset Submission
*****


### Eloisa Moreira de Lira
### Geologist Bach and Reservoir Geophysist
eloisalira@gmail.com

**This assigmnet is destinated to conclude the IBM Advanced Data Science Specialization. All these informations came from free datasets referenced bellow. THIS RESERCH HAVE ONLY EDUCATIONAL PURPOSE.**

### Geological Overview
**********

The Otway Basin is a north-west to south-east striking, divergent margin, rift and drift basin. It is approximately 500km long from Cape Jaffa in South Australia to north-west Tasmania and forms part of the 4 000km long Jurassic-Cretaceous Australian Southern Rift System.[1]

The basin dates from the late Jurassic to late Cretaceous periods and formed by multi-stage rifting during the breakup of Gondwana and the separation of the Antarctic and Australian plates. The basin contains a significant amount of natural gas and is a current source of commercial extraction.[5]

## Summary Facts

In [1]:
import folium

S_australia=folium.Map(location= [-38, 141], zoom_start=4, tiles='Stamen Terrain') #from Google Maps

folium.Marker(
    location=[-38, 141],
    popup='Otway Basin',
    icon=folium.Icon(color='Red')
).add_to(S_australia)

S_australia

|Summary		                        |                                                   |
|-----------------------------------	|-------------------------------------------------	|
| Age                               	| Jurassic - Late Cretaceous                      	|
| Onshore Exploration Well Density  	| 1 well per 121 km2                              	|
| Onshore success ratio             	| 12,5%                                           	|
| Offshore exploration well Density 	| 1 well per 5,000 km2                            	|
| Depth to target zone              	| 1000-4000+ m                                    	|
| Thickness                         	| >9 km                                           	|
| First commercial discovery        	| 1987 (Katnook 1) 1967 CO2 (Caroline 1)          	|
| Identified reserves               	| 78.1 PJ (74 x 1012 Btu) sales gas (in 6 fields) 	|


Source data[4]



![Otway Basin 2 ](https://www.ga.gov.au/__data/assets/image/0013/13045/9167-65.jpg)

**Location of Basin in yellow-greenish.**

The Late Jurassic-Cenozoic Otway Basin is a large, northwest trending on/offshore basin on the southern Australian passive margin. Exploration is mature onshore and immature offshore, with >200 wells in South Australia, Victoria and Tasmania. Commercial gas discoveries include industrial grade CO2. No commercial oil discoveries have been identified.

The basin formed by multi-stage rift-sag and inversion phases. Late Jurassic to Early Cretaceous rifting resulted in the east-west trending Inner Otway Basin. Late Cretaceous rifting, culminating in continental breakup in the Maastrichtian, produced northwest-southeast trending depocentres beneath the outer shelf and slope. Multiple phases of compression in the Cretaceous-Recent resulted in inversion and wrenching of pre-existing structures.

The basin contains five major depocentres, the mainly onshore Inner Otway Basin, the offshore Morum, Nelson and Hunter Sub-basins and eastern Torquay Sub-basin. The Latest Jurassic-Early Cretaceous Otway Supergroup comprises up to 8km of continental and fluvio-lacustrine sediments that accumulated in grabens and half-grabens of the first rifting event. Coastal-plain, deltaic and marine sediments of the Late Cretaceous Sherbrook Group are up to 5km thick. The Paleocene-middle Eocene Wangerrip Group sediments were deposited in coastal plain, deltaic and inner shelf settings and are separated from the open marine, mixed carbonates/siliciclastics of the Eocene-Miocene Nirrandra and Heytesbury groups, by a major unconformity.

The main exploration targets in the Otway Basin are the Waarre Sandstone at the base of the Sherbrook Group and sandstones of the Pretty Hill Formation and Katnook Sandstone/Windermere Sandstone Member in the Early Cretaceous section. The main source rocks occur in the Early Cretaceous section. Regional and intraformational seals exist in the Pretty Hill, Laira, Eumeralla and Flaxman formations, the Belfast, Skull Creek and Pember mudstones and mudstones and marls of the Wangerrip, Nirrandra and Heytesbury groups. Play types include faulted anticlines, large anticlinal features and tilted fault blocks.[1]



![Stratigraphic Chart](https://www.ga.gov.au/__data/assets/image/0016/13048/GA18588.jpg)

#### Stratigraphic and Basin Event Chart of the Otway Basin

### Structural Elements

The basin includes five significant depocenters. The Inner Otway Basin, Torquay Sub-basin, Morum Sub-basin, Nelson Sub-basin, and Hunter Sub-basin formed as a result of two major, basin-wide rifting phases. At the onset of major north-south rifting in the late Jurassic several east-west trending extensional depocenters developed in the onshore part of the basin to define the Inner Otway Basin. Renewed rifting in the late Cretaceous was driven by a change in crustal extension style from north-south to northeast-southwest resulting in the structurally different, northwest-southeast trending Torquay, Morum, Nelson, and Hunter Sub-basins in the mid- and off-shore. Following each rifting episode the basin underwent compressional phases resulting in the inversion and wrenching of pre-existing structures.[2]

![Structural](https://www.ga.gov.au/__data/assets/image/0014/13046/GA7391.gif)

### Loading Wells

Referência de download

https://sarigbasis.pir.sa.gov.au/WebtopEw/ws/samref/sarig1/wcir/Record;jsessionid=E0E64FB23D26AC838007680E46C88B18

https://towardsdatascience.com/10-steps-in-pandas-to-process-las-file-and-plot-610732093338 (mais sucinto)

### condicionamento dos dados segue tais instruções

https://prd-wret.s3-us-west-2.amazonaws.com/assets/palladium/production/s3fs-public/atoms/files/LAS%20QA-QC%20Utilities.pdf

1 Reading File
2 Data Inspection
3 Column Selection
4 Missing Data
5 Dropping Rows
6 Statistics
7 Filtering
8 Adding Column
9 Indexing
10 Plotting

### instalando bibliotecas e dependencias

In [3]:
#pip install --upgrade pip
# to update pip

Requirement already up-to-date: pip in /home/familia/anaconda3/lib/python3.7/site-packages (20.1.1)
Note: you may need to restart the kernel to use updated packages.


In [2]:
!pip install lasio
!pip install -r optional-packages.txt
!pip install --upgrade lasio
import matplotlib.pyplot as plt


ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'optional-packages.txt'
Requirement already up-to-date: lasio in /home/familia/anaconda3/lib/python3.7/site-packages (0.25.1)


In [3]:
!pip install welly

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import lasio
import sys
%matplotlib inline

print("installed")

installed


In [5]:
# POCOS DO CAMPO DE Otway

P1=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Balnaves 001/Balnaves 1.las")

P2=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Viewbank 001/Viewbank 1.las")

P3=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Wynn 002/Wynn 2.las")      

P4=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/McNamara Park 001/McNamara Park 1.las")

P5=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Northumberland 002/Northumberland 2.las")

P6=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Banyula 001/Banyula 1.las")

P7=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Bool Lagoon 001/Bool Lagoon 1.las")

P8=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Reedy Creek 001/Reedy Creek 1.las")

P9=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Patrick 001/Patrick 1.las")

P10=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Camelback 001/Camelback 1.las")

P11=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Lake Hawdon 001/Lake Hawdon 1.las")

P12=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Crankshaft 001/Crankshaft 1.las")

P13=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Laira 001/Laira 1.las")

P14=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Killanoola Southeast 001/Killanoola Southeast 1.las")

P15=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Katnook 004/Katnook 4.las")

P16=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Jacaranda Ridge 001/Jacaranda Ridge 1.las")

P17=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Penley 001/Penley 1.las")

P18=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Cowrie 001/Cowrie 1.las")

P19=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Hollick 001/Hollick 1.las")

P20=lasio.read("/home/familia/Desktop/Pioneer_Field_LAS/Untitled Folder/OtwayBasin/Bungaloo 001/Bungaloo 1.las")

In [6]:
#examinando as curvas
P17.curves


[CurveItem(mnemonic=DEPTH, unit=M, value=, descr=Depth, original_mnemonic=DEPTH, data.shape=(12511,)),
 CurveItem(mnemonic=CALI, unit=in, value=, descr=Caliper         HCAL Edited, penley_1.dlis, original_mnemonic=CALI, data.shape=(12511,)),
 CurveItem(mnemonic=DRHO, unit=g/cm3, value=, descr=DensityCorr     HDRA Edited, penley_1.dlis, original_mnemonic=DRHO, data.shape=(12511,)),
 CurveItem(mnemonic=DT, unit=us/ft, value=, descr=Sonic           DT Edited, penley_1.dlis, original_mnemonic=DT, data.shape=(12511,)),
 CurveItem(mnemonic=GR, unit=gAPI, value=, descr=GammaRay        GR Spliced, penley_1.dlis, original_mnemonic=GR, data.shape=(12511,)),
 CurveItem(mnemonic=MINV, unit=ohmm, value=, descr=MicroInv        HMIN Edited, penley_1.dlis, original_mnemonic=MINV, data.shape=(12511,)),
 CurveItem(mnemonic=MNOR, unit=ohmm, value=, descr=MicroNor        HMNO Edited, penley_1.dlis, original_mnemonic=MNOR, data.shape=(12511,)),
 CurveItem(mnemonic=NPHI, unit=dec, value=, descr=Neutron     

In [7]:
P8.write(sys.stdout)

~Version ---------------------------------------------------
VERS.            2.0 : CWLS log ASCII Standard -VERSION 2.0
WRAP.             NO : SINGLE LINE PER DEPTH STEP
CREA. 22-Apr-17 8:59 : 13 AM
~Well ------------------------------------------------------
STRT .M            0.0 : START DEPTH
STOP .M      2671.1148 : STOP DEPTH
STEP .M         0.1524 : STEP VALUE
NULL .         -999.25 : NULL VALUE
DATE .       22-Apr-17 : LAS file Creation Date
WELL .   Reedy Creek 1 : Well Name
UWI  . Reedy Creek 001 : Unique Well Identifier (UWI)
COMP .    Gas And Fuel : Company
STATE.              SA : State
COUNT.       AUSTRALIA : Country
API  .            1772 : API Number
LOC  .           Otway : Location
GDAT .   GDA94 Zone 54 : Geodetic Datum
LATI .       5869174.9 : Latitude/Northing
LONG .       425252.24 : Longitude/Easting
EKB  .            17.2 : KB Elevation
EGL  .            12.2 : GL Elevation
TDEP .          2669.0 : Total Depth
DATS .       27-Feb-92 : Spud Date
DATR .       28-

  245.82120    -999.25    -999.25    -999.25   62.36750    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25
  245.97360    -999.25    -999.25    -999.25   60.13480    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25
  246.12600    -999.25    -999.25    -999.25   56.87740    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25
  246.27840    -999.25    -999.25    -999.25   57.38980    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25
  246.43080    -999.25    -999.25    -999.25   61.78190    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25
  246.58320    -999.25    -999.25    -999.25   63.02630    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25    -999.25
  246.73560    -999.25    -999.25    -999.25   64.38050    -999.25    

  543.15360   16.31020    0.02660  169.47070   79.31360    -999.25    -999.25    -999.25    -999.25   13.66980    2.24340    4.71870    -999.25   19.33470
  543.30600   16.41770    0.02260  172.05010   81.03380    -999.25    -999.25    -999.25    -999.25   14.46630    2.18740    4.85560    -999.25   19.26540
  543.45840   16.32810    0.01830  174.43350   81.18020    -999.25    -999.25    -999.25    -999.25   14.63840    2.15060    4.90490    -999.25   19.74170
  543.61080   16.20950    0.02080  179.50460   80.63120    -999.25    -999.25    -999.25    -999.25   14.82390    2.13330    5.01800    -999.25   20.95450
  543.76320   16.23960    0.02360  181.25050   79.89920    -999.25    -999.25    -999.25    -999.25   15.35640    2.14540    5.15330    -999.25   20.90280
  543.91560   16.72990    0.02410  181.70890   80.44820    -999.25    -999.25    -999.25    -999.25   17.75280    2.15400    5.63410    -999.25   20.59790
  544.06800   17.57970    0.02630  181.45050   81.65600    -999.25    

  725.72880   13.52660    0.01460  134.36490  117.85400    -999.25    -999.25    -999.25    -999.25    1.39610    2.14610    1.27270    -999.25  -38.29450
  725.88120   13.61290    0.01690  134.70660  118.03700    -999.25    -999.25    -999.25    -999.25    1.40380    2.13040    1.27830    -999.25  -38.26280
  726.03360   13.41440    0.01620  132.13980  114.70640    -999.25    -999.25    -999.25    -999.25    1.38620    2.12550    1.25960    -999.25  -39.85960
  726.18600   13.31090    0.01100  131.14390  111.77830    -999.25    -999.25    -999.25    -999.25    1.37670    2.12510    1.24970    -999.25  -40.01770
  726.33840   13.41870    0.00940  132.11470  112.18090    -999.25    -999.25    -999.25    -999.25    1.35100    2.13400    1.22760    -999.25  -41.85480
  726.49080   13.50900    0.00940  132.91480  114.66980    -999.25    -999.25    -999.25    -999.25    1.34540    2.16020    1.22180    -999.25  -42.43590
  726.64320   13.61690    0.01050  134.46900  116.09720    -999.25    

 1092.55560   12.24160    0.01700  114.74720  119.35460    -999.25    -999.25    -999.25    -999.25    1.36780    2.22910    1.29690    -999.25  -41.16560
 1092.70800   12.20710    0.02110  115.28060  123.56370    -999.25    -999.25    -999.25    -999.25    1.29210    2.20710    1.22930    -999.25  -41.36790
 1092.86040   12.18350    0.02190  115.94320  128.61460    -999.25    -999.25    -999.25    -999.25    1.15870    2.19940    1.11050    -999.25  -41.10960
 1093.01280   12.17920    0.01930  115.99310  126.34540    -999.25    -999.25    -999.25    -999.25    1.11970    2.18490    1.07630    -999.25  -41.64300
 1093.16520   12.17520    0.01720  115.38900  116.39000    -999.25    -999.25    -999.25    -999.25    1.09420    2.17940    1.05370    -999.25  -40.74920
 1093.31760   12.16650    0.01390  114.72640  113.79140    -999.25    -999.25    -999.25    -999.25    1.09780    2.17230    1.05670    -999.25  -40.85720
 1093.47000   12.13190    0.01290  112.61800  115.62140    -999.25    

 1413.35760   12.53050    0.00950   94.31710  142.23000    -999.25    -999.25    -999.25    -999.25    2.70990    2.40530    2.82170    -999.25  -49.36420
 1413.51000   12.53220    0.00720   93.66710  149.98930    -999.25    -999.25    -999.25    -999.25    2.71310    2.40330    2.80970    -999.25  -49.25480
 1413.66240   12.53240    0.00630   93.35040  151.19720    -999.25    -999.25    -999.25    -999.25    2.73480    2.40070    2.82440    -999.25  -49.60700
 1413.81480   12.53040    0.00490   93.08370  146.14630    -999.25    -999.25    -999.25    -999.25    2.76690    2.39530    2.84770    -999.25  -49.73170
 1413.96720   12.53240    0.00500   93.16280  141.09540    -999.25    -999.25    -999.25    -999.25    2.76420    2.39660    2.84040    -999.25  -49.81810
 1414.11960   12.56620    0.00900   93.81290  131.46950    -999.25    -999.25    -999.25    -999.25    2.68870    2.41260    2.76280    -999.25  -49.73020
 1414.27200   12.55700    0.01110   94.30050  129.82240    -999.25    

 1746.19920    9.03770    0.03030   84.57500  122.59970    2.14740    1.97720    0.28320    2.62160    3.17250    2.48830    3.11610    3.14700  134.76280
 1746.35160    8.93980    0.02910   84.43330  113.50170    1.54610    1.55710    0.27030    2.52520    2.64730    2.48300    2.62730    2.48440  134.90020
 1746.50400    8.77900    0.02000   84.61670   95.26840    0.72790    0.97800    0.22500    2.36890    1.91850    2.42920    1.96230    1.21370  134.97370
 1746.65640    8.76360    0.01470   84.64990   87.58720    0.78590    0.99640    0.20490    2.34730    1.74250    2.39220    1.77340    0.96140  134.69790
 1746.80880    8.76360    0.00440   84.50830   79.23490    0.77700    0.97610    0.19120    2.30830    1.63800    2.32520    1.67950    0.90390  134.21370
 1746.96120    8.76840    0.00430   83.89580   80.05530    0.76650    0.96820    0.20110    2.37350    1.82470    2.30610    1.86090    0.92740  134.32120
 1747.11360    8.78250    0.00650   83.69160   83.44840    0.87670    

 2129.79000    9.30810    0.00700   78.22470  137.92470    4.85830    4.81880    0.15840    2.73440    5.39190    2.52200    5.79810    5.84580  107.40220
 2129.94240    9.34170    0.01170   79.60390  148.32780    3.92810    3.66510    0.16790    2.74790    4.99800    2.52870    5.37940    4.78420  106.80270
 2130.09480    9.36120    0.01310   80.23310  149.33460    3.66900    3.40020    0.18130    2.73540    4.85570    2.52780    5.20160    4.47920  106.69610
 2130.24720    9.37890    0.01370   80.19560  139.37890    3.35850    3.28050    0.20440    2.72760    4.97350    2.51880    5.24620    4.99040  106.78810
 2130.39960    9.37890    0.01270   79.64550  133.67400    3.37090    3.29420    0.20620    2.76150    5.19670    2.51280    5.46270    4.99120  106.84030
 2130.55200    9.37890    0.01190   78.64550  125.62000    2.83020    3.18600    0.19510    2.89300    5.56940    2.50940    5.89660    5.36530  106.74210
 2130.70440    9.38060    0.01160   78.35380  121.36930    2.76780    

 2545.08000    8.75480    0.02190   82.20400   83.41110    0.80000    0.91590    0.18220    2.67470    2.70760    2.33110    2.35920    1.65540  111.67500
 2545.23240    8.71470    0.03100   82.45410   86.54320    0.80930    0.99380    0.19000    2.35220    2.46410    2.37450    2.16290    1.61890  111.01870
 2545.38480    8.70530    0.03310   83.07490   89.74990    0.77820    0.99640    0.20260    2.24810    2.20640    2.38540    1.95230    1.50920  111.28940
 2545.53720    8.70000    0.02900   84.11240   94.00060    0.66070    0.90210    0.22540    2.27880    1.75130    2.36260    1.60100    0.82290  110.67030
 2545.68960    8.70000    0.02300   84.13750   95.52940    0.59740    0.79380    0.22620    2.30110    1.64670    2.33280    1.52610    0.73390  111.03820
 2545.84200    8.70000    0.01190   82.89150   98.40050    0.58680    0.77670    0.20250    2.31320    1.73580    2.28010    1.62710    0.73190  111.31250
 2545.99440    8.70000    0.01550   79.95390  103.91900    0.64310    

### vamos transformar como pandas data frame

In [8]:
### teste com alguns pocos para colocar no folium com as localizações

poco=pd.DataFrame()
poco['x']=[473859.92,478652.81,489424.08,466607.57,470961.25,453653.71,467468.47]
poco['y']=[5855457.54,5869002.58,5859481.37,5813084.45,5789741.44,5865409.47,5888985.98]
poco['name']=['P1','P2','P3','P4','P5','P6','P7']
poco['real name']=['Balnaves 1','Viewbank 1','Wynn 2','McNamara Park 1','Northumberland 2','Banyula 001','Bool Lagoon 1']
poco.head()

x           y name         real name
0  473859.92  5855457.54   P1        Balnaves 1
1  478652.81  5869002.58   P2        Viewbank 1
2  489424.08  5859481.37   P3            Wynn 2
3  466607.57  5813084.45   P4   McNamara Park 1
4  470961.25  5789741.44   P5  Northumberland 2

In [9]:
S_australia=folium.Map(location= [-38, 141], zoom_start=4, tiles='Stamen Terrain') #from Google Maps

folium.Marker(
    location=[-38, 141],
    popup='Otway Basin',
    icon=folium.Icon(color='Red')
).add_to(S_australia)

folium.Marker(location=[poco['x'], poco['y']], poupup=poco['name'], icon=folium.Icon(color='green') ).add_to(S_australia)

S_australia

NameError: name 'folium' is not defined

In [19]:
P1=P1.df()
P1['Well']='P1'

P2=P2.df()
P2['Well']='P2'

P3=P3.df()
P3['Well']='P3'

P4=P4.df()
P4['Well']='P4'

P5=P5.df()
P5['Well']='P5'

P6=P6.df()
P6['Well']='P6'


P7=P7.df()
P7['Well']='P7'


P8=P8.df()
P8['Well']='P8'


P9=P9.df()
P9['Well']='P9'


P10=P10.df()
P10['Well']='P10'


P11=P11.df()
P11['Well']='P11'


P12=P12.df()
P12['Well']='P12'


P13=P13.df()
P13['Well']='P13'


P14=P14.df()
P14['Well']='P14'


P15=P15.df()
P15['Well']='P15'


P16=P16.df()
P16['Well']='P16'


P18=P18.df()
P18['Well']='P18'


P19=P19.df()
P19['Well']='P19'


P20=P20.df()
P20['Well']='P20'



In [20]:
#print(df_P_25_28.columns)
print('P1', P1.columns)
print('P2',P2.columns)
print('P3',P2.columns)
print('P4',P2.columns)

P1 Index(['CALI', 'DRHO', 'DT', 'GR', 'MINV', 'MNOR', 'NPHI', 'PEF', 'RDEP',
       'RHOB', 'RMED', 'RMIC', 'SP', 'Well'],
      dtype='object')
P2 Index(['CALI', 'DRHO', 'DT', 'GR', 'NPHI', 'PEF', 'RDEP', 'RHOB', 'RMED',
       'RMIC', 'SP', 'Well'],
      dtype='object')
P3 Index(['CALI', 'DRHO', 'DT', 'GR', 'NPHI', 'PEF', 'RDEP', 'RHOB', 'RMED',
       'RMIC', 'SP', 'Well'],
      dtype='object')
P4 Index(['CALI', 'DRHO', 'DT', 'GR', 'NPHI', 'PEF', 'RDEP', 'RHOB', 'RMED',
       'RMIC', 'SP', 'Well'],
      dtype='object')


 #pocos com os perfis ja bem setados
    
    
    CALI','DT','GR','NPHI','RHOB','SP','Well


In [21]:
P1_trat=P1[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P2_trat=P2[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P3_trat=P3[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P4_trat=P4[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P5_trat=P5[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P6_trat=P6[['CALI','DT','GR','NPHI','RHOB','SP','Well']]

In [22]:
P7_trat=P7[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P8_trat=P8[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P9_trat=P9[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P10_trat=P10[['CALI','DT','GR','NPHI','RHOB','SP','Well']]

In [23]:
P11_trat=P11[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P12_trat=P12[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P13_trat=P13[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P14_trat=P14[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P15_trat=P15[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P16_trat=P16[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
#P17_trat=P17[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P18_trat=P18[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P19_trat=P19[['CALI','DT','GR','NPHI','RHOB','SP','Well']]
P20_trat=P20[['CALI','DT','GR','NPHI','RHOB','SP','Well']]

### juntando todos os data frames

In [24]:
result1 = P1_trat.append([P2_trat, P3_trat,P4_trat,P5_trat])
result2=result1.append([P6_trat,P7_trat,P8_trat,P9_trat])
result3=result2.append([P10_trat,P11_trat,P12_trat, P13_trat])
result4=result3.append([P14_trat,P15_trat,P16_trat,P18_trat])
all_wells=result4.append([P19_trat,P20_trat])     

all_wells.tail(10)

CALI  DT  GR  NPHI  RHOB  SP Well
DEPTH                                       
3718.7124   NaN NaN NaN   NaN   NaN NaN  P20
3718.8648   NaN NaN NaN   NaN   NaN NaN  P20
3719.0172   NaN NaN NaN   NaN   NaN NaN  P20
3719.1696   NaN NaN NaN   NaN   NaN NaN  P20
3719.3220   NaN NaN NaN   NaN   NaN NaN  P20
3719.4744   NaN NaN NaN   NaN   NaN NaN  P20
3719.6268   NaN NaN NaN   NaN   NaN NaN  P20
3719.7792   NaN NaN NaN   NaN   NaN NaN  P20
3719.9316   NaN NaN NaN   NaN   NaN NaN  P20
3720.0840   NaN NaN NaN   NaN   NaN NaN  P20

### Missing values
are very common in LAS files because depth of interest may vary for specific measurements. Printed values of -999.2500 in LAS files means null value. Missing values are not common in the middle part of the dataset. Usually, it happens in the head or tail of a file. isna function returns True if the location has a null value, otherwise, it will be False. We may add up these boolean values using sum() function as below:

In [25]:
all_wells.isna().sum()


CALI     84928
DT       84872
GR       10303
NPHI    130854
RHOB    126030
SP       96333
Well         0
dtype: int64

Let’s suppose that we want to preserve GR, DT, and SPOR as important logs in the data set. So, to get rid of missing values we can drop rows that one of those logs is missed in that row.

### dropping Rows

This line of code will drop all rows that one (or more) of the subset logs (‘GR’, ‘DT’, ‘SPOR’) has a null value.

In [26]:
all_wells_dropped = all_wells.dropna(subset=['CALI','DT','GR','NPHI','RHOB','SP'],axis=0, how='any')

### Statistics

Run describe command to see data statistics. This can be helpful to see the data range and outliers.

In [27]:
all_wells_dropped.describe()


CALI             DT             GR           NPHI  \
count  181552.000000  181552.000000  181552.000000  181552.000000   
mean        8.932678      92.280629     102.474539       0.249096   
std         1.230664      18.271113      26.576768       0.108896   
min         0.000000       0.000000       0.000000      -0.005500   
25%         8.407100      79.515300      86.977950       0.193900   
50%         8.735500      87.875000     103.508800       0.256500   
75%         9.289400     101.300000     120.753750       0.318500   
max        23.612400    1183.000000     261.568600       0.871500   

                RHOB             SP  
count  181552.000000  181552.000000  
mean        2.374567      21.927954  
std         0.181838      83.363314  
min         0.000000    -135.375000  
25%         2.271500     -28.198650  
50%         2.405700       0.775350  
75%         2.507200      67.189550  
max         3.192400    4528.000000

**It's important to understand that there are many spurious values meanwhile well logging. To avoid that all the values will be filtered acording the ranges. Consult the referencce [5] to understant the values.**

In [28]:
all_wells_dropped= all_wells_dropped[(all_wells_dropped.NPHI > 0) & (all_wells_dropped.NPHI  <= 60)]
all_wells_dropped= all_wells_dropped[(all_wells_dropped.SP > -160)&(all_wells_dropped.GR <= 40)]
all_wells_dropped= all_wells_dropped[(all_wells_dropped.GR > 0) & (all_wells_dropped.GR  <= 200)]
all_wells_dropped= all_wells_dropped[(all_wells_dropped.RHOB> 1) & (all_wells_dropped.RHOB<= 3)]
all_wells_dropped= all_wells_dropped[(all_wells_dropped.DT > 0) & (all_wells_dropped.DT <= 200)]

Petrophysical properties of rocks have a normal range. The readings out of that range should not be credited for validity. These ranges are:

### Normalizing data

In [29]:
from sklearn import preprocessing

In [30]:
# all_wells_dropped_norm=all_wells_dropped
all_wells_dropped_norm=all_wells_dropped[['DT','GR','NPHI','RHOB','SP']].values


#all_wells_dropped[['DT','GR','NPHI','RHOB','SP']] = all_wells_dropped_norm[['DT','GR','NPHI','RHOB','SP']]
print(all_wells_dropped_norm)
print("Let's slice the data frame because has a string value!")

[[ 7.851890e+01  3.863670e+01  9.130000e-02  2.419400e+00 -7.252080e+01]
 [ 8.053390e+01  3.869380e+01  1.143000e-01  2.374900e+00 -7.193750e+01]
 [ 8.107420e+01  3.847380e+01  1.363000e-01  2.346300e+00 -7.154170e+01]
 ...
 [ 7.350590e+01  3.817110e+01  9.900000e-02  2.420900e+00  1.853658e+02]
 [ 7.296150e+01  3.805800e+01  9.690000e-02  2.439500e+00  1.845434e+02]
 [ 7.325270e+01  3.794480e+01  9.480000e-02  2.458800e+00  1.837211e+02]]
Let's slice the data frame because has a string value!


In [31]:
min_max_scaler = preprocessing.MinMaxScaler()
pocos_scaled = min_max_scaler.fit_transform(all_wells_dropped_norm)
all_wells_dropped_normalizado=pd.DataFrame(pocos_scaled)
# print(all_wells_dropped_normalizado)

X = all_wells_dropped

print(all_wells_dropped_normalizado)

print(all_wells_dropped_normalizado.dtypes)

X.replace({'DT': all_wells_dropped_normalizado[0]})
print(X)


             0         1         2         3         4
0     0.231318  0.965801  0.120289  0.796884  0.097580
1     0.254066  0.967234  0.151029  0.765791  0.099388
2     0.260166  0.961715  0.180433  0.745808  0.100615
3     0.245723  0.970412  0.204491  0.741126  0.102617
4     0.267810  0.996977  0.213713  0.721702  0.111076
5     0.267112  0.932041  0.212911  0.727012  0.109527
6     0.266524  0.948216  0.208901  0.730715  0.107202
7     0.272441  0.994720  0.196739  0.729178  0.105846
8     0.195084  0.958411  0.000000  0.820570  0.368877
9     0.203563  0.990054  0.000000  0.824483  0.369210
10    0.544999  0.957408  0.403903  0.676006  0.271425
11    0.539355  0.916541  0.398022  0.678871  0.269221
12    0.540201  0.880366  0.336407  0.685788  0.260514
13    0.549515  0.877807  0.309810  0.681666  0.257079
14    0.588651  0.904744  0.312216  0.660075  0.259231
15    0.608596  0.952007  0.370489  0.648966  0.261531
16    0.684848  0.974962  0.527265  0.620668  0.262496
17    0.67

ValueError: Replacement not allowed with overlapping keys and values

In [32]:
all_wells_dropped_norm=all_wells_dropped_norm.values

AttributeError: 'numpy.ndarray' object has no attribute 'values'

In [14]:
min_max_scaler = preprocessing.MinMaxScaler()
print("printou")

printou


In [15]:
pocos_scaled = min_max_scaler.fit_transform(all_wells_dropped_norm)


NameError: name 'all_wells_dropped_norm' is not defined

In [16]:
all_wells_dropped_normalizado=pd.DataFrame(pocos_scaled)
all_wells_dropped_normalizado=all_wells_dropped_normalizado.header(['DT','GR','NPHI','RHOB','SP'])


NameError: name 'pocos_scaled' is not defined

In [17]:
all_wells_dropped_normalizado.header()

NameError: name 'all_wells_dropped_normalizado' is not defined

In [18]:
|Perfis Normalizados                          |
|---------------------------------------------|
|Antes                 |        Depois      |

SyntaxError: invalid syntax (<ipython-input-18-844d5720858b>, line 1)

### reindex the depth

__lasio__ converts the depth column as an index when we read the LAS file. We can call depth as a new column for plotting aim and reindex dataset.

In [25]:
all_wells_dropped = all_wells_dropped.rename_axis('Depth').reset_index()
all_wells_dropped.tail(22)

ValueError: cannot insert Depth, already exists

## Ploting logs

In [ ]:
def log_plot(logs):
    logs = logs.sort_values(by='Depth')
    top = logs.Depth.min()
    bot = logs.Depth.max()
    
    f, ax = plt.subplots(nrows=1, ncols=6, figsize=(12,8))
    ax[0].plot(logs.CALI, logs.Depth, color='green')
    ax[1].plot(logs.GR, logs.Depth, color='red')
    ax[2].plot(logs.NPHI, logs.Depth, color='black')
    ax[3].plot(logs.SP, logs.Depth, color='blue')
    ax[4].plot(logs.RHOB, logs.Depth, color='c')
    ax[5].plot(logs.DT, logs.Depth, color='m')
        
    for i in range(len(ax)):
        ax[i].set_ylim(top,bot)
        ax[i].invert_yaxis()
        ax[i].grid()
        
    ax[0].set_xlabel("CALI")
    ax[0].set_xlim(logs.CALI.min(),logs.CALI.max())
    ax[0].set_ylabel("Depth(ft)")
    ax[1].set_xlabel("GR")
    ax[1].set_xlim(logs.GR.min(),logs.GR.max())
    ax[2].set_xlabel("NPHI")
    ax[2].set_xlim(logs.NPHI.min(),logs.NPHI.max())
    ax[3].set_xlabel("SP")
    ax[3].set_xlim(logs.SP.min(),logs.SP.max())
    ax[4].set_xlabel("RHOB")
    ax[4].set_xlim(logs.RHOB.min(),logs.RHOB.max())
    ax[5].set_xlabel("DT")
    ax[5].set_xlim(logs.DT.min(),logs.DT.max())
   
    
    ax[1].set_yticklabels([]); ax[2].set_yticklabels([]);
    ax[3].set_yticklabels([])
    ax[4].set_yticklabels([]); ax[5].set_yticklabels([]) 
    
    f.suptitle('Wells', fontsize=14,y=0.94)

In [ ]:
log_plot(all_wells_dropped)

## References

[1]https://www.ga.gov.au/scientific-topics/energy/province-sedimentary-basin-geology/petroleum/offshore-southern-australia/otway , visited in Jun/2020.

[2]https://en.wikipedia.org/wiki/Otway_Basin, visited in Jun/2020.

[3]https://www.spec2000.net, visited in Jun/2020.

[4]http://www.energymining.sa.gov.au/petroleum/prospectivity/otway_basin, visited in Jun/2020.

[5]https://readthedocs.org/projects/lasio/downloads/pdf/latest/

[6]https://wiki.aapg.org/Well_log_analysis_for_reservoir_characterization 